gitpod-io/gitpod
# [ws-manager] Refactor connection to ws-daemon

## Description

- Refactor connection to ws-daemon removing calls to the k8s API using an informer that filters `ws-daemon` pods+
- Replace custom grpc pool.

## How to test

- Check ws-manager logs and check the grpc pool can reach ws-daemon

``` 
{"component":"grpc","level":"warning","message":"2022/07/29 23:49:45 WARNING: [core] grpc: addrConn.createTransport failed to connect to {image-builder-mk3.default.svc.cluster.local:8080 image-builder-mk3.default.svc.cluster.local:8080 \u003cnil\u003e \u003cnil\u003e 0 \u003cnil\u003e}. Err: connection error: desc = \"transport: Error while dialing dial tcp: lookup image-builder-mk3.default.svc.cluster.local on 10.30.0.10:53: read udp 192.168.65.211:55593-\u003e10.30.0.10:53: read: connection refused\"","serviceContext":{"service":"ws-manager","version":"commit-fb5c172918f5aa4dbaaf90c5eb988364f2759c64"},"severity":"WARNING","time":"2022-07-29T23:49:45Z"}
2022/07/29 23:49:52 new pool success: address:192.168.65.216, index:0, current:8, ref:0. option:{0x1797480 8 64 64 false}
``` 


- Delete ws-daemon
- Check the grpc pool removes the IP address and after some time adds a new one

```
2022/07/29 23:51:10 close pool success: address:192.168.65.216, index:0, current:0, ref:0. option:{0x1797480 8 64 64 false}
{"component":"grpc","level":"warning","message":"2022/07/29 23:51:12 WARNING: [core] grpc: addrConn.createTransport failed to connect to {192.168.65.229:8080 wsdaemon \u003cnil\u003e \u003cnil\u003e 0 \u003cnil\u003e}. Err: connection error: desc = \"transport: Error while dialing dial tcp 192.168.65.229:8080: connect: connection refused\"","serviceContext":{"service":"ws-manager","version":"commit-fb5c172918f5aa4dbaaf90c5eb988364f2759c64"},"severity":"WARNING","time":"2022-07-29T23:51:12Z"}
{"component":"grpc","level":"warning","message":"2022/07/29 23:51:13 WARNING: [core] grpc: addrConn.createTransport failed to connect to {192.168.65.229:8080 wsdaemon \u003cnil\u003e \u003cnil\u003e 0 \u003cnil\u003e}. Err: connection error: desc = \"transport: Error while dialing dial tcp 192.168.65.229:8080: connect: connection refused\"","serviceContext":{"service":"ws-manager","version":"commit-fb5c172918f5aa4dbaaf90c5eb988364f2759c64"},"severity":"WARNING","time":"2022-07-29T23:51:13Z"}
{"component":"grpc","level":"warning","message":"2022/07/29 23:51:14 WARNING: [core] grpc: addrConn.createTransport failed to connect to {192.168.65.229:8080 wsdaemon \u003cnil\u003e \u003cnil\u003e 0 \u003cnil\u003e}. Err: connection error: desc = \"transport: Error while dialing dial tcp 192.168.65.229:8080: connect: connection refused\"","serviceContext":{"service":"ws-manager","version":"commit-fb5c172918f5aa4dbaaf90c5eb988364f2759c64"},"severity":"WARNING","time":"2022-07-29T23:51:14Z"}
2022/07/29 23:51:17 new pool success: address:192.168.65.229, index:0, current:8, ref:0. option:{0x1797480 8 64 64 false}
``` 

## Release Notes

```release-note
NONE
```



## Werft options:
